modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -1,66 +1,164 @@
+      def pullLatestSnapshot: F[LatestSnapshotTuple] =
+        maybeMajorityPeerIds.fold(pullLatestSnapshotFromRandomPeer)(pullLatestSnapshotWithMajorityHash)
 
       def pullGlobalSnapshot(hash: Hash): F[Option[Hashed[GlobalIncrementalSnapshot]]] =
         pullGlobalSnapshot(l0GlobalSnapshotClient.get(hash)).handleErrorWith { e =>
           logger
-            .warn(e)(s"Failure pulling single snapshot with hash=$hash")
-            .map(_ => none[Hashed[GlobalIncrementalSnapshot]])
+            .warn(e)(s"Failure pulling single snapshot with hash=${hash.show}")
+            .as(none)
         }
 
       def pullGlobalSnapshot(ordinal: SnapshotOrdinal): F[Option[Hashed[GlobalIncrementalSnapshot]]] =
         pullGlobalSnapshot(l0GlobalSnapshotClient.get(ordinal)).handleErrorWith { e =>
           logger
-            .warn(e)(s"Failure pulling single snapshot with ordinal=$ordinal")
-            .map(_ => none[Hashed[GlobalIncrementalSnapshot]])
+            .warn(e)(s"Failure pulling single snapshot with ordinal=${ordinal.show}")
+            .as(none)
+        }
+
+      def pullGlobalSnapshots: F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        maybeMajorityPeerIds
+          .fold(pullGlobalSnapshotsFromRandomPeer)(pullGlobalSnapshotsFromMajority)
+          .handleErrorWith { e =>
+            logger
+              .warn(e)(s"Failure pulling global snapshots")
+              .as(
+                List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[LatestSnapshotTuple]
+              )
+          }
+
+      private def pullGlobalSnapshotsFromMajority(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        lastGlobalSnapshotStorage.getOrdinal.flatMap {
+          _.fold {
+            pullLatestSnapshotWithMajorityHash(majorityPeerIds).map(_.asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+          } { lastStoredOrdinal =>
+            for {
+              msd <- getLatestMajoritySnapshotData(majorityPeerIds)
+              l0Peers <- globalL0ClusterStorage.getPeers
+              pulled <- pullVerifiedSnapshots(lastStoredOrdinal.next, msd, l0Peers)
+            } yield pulled.asRight[LatestSnapshotTuple]
+          }
+        }
+
+      private def pullLatestSnapshotWithMajorityHash(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)] =
+        globalL0ClusterStorage.getPeers
+          .map(nes => Random.shuffle(nes.toSortedSet))
+          .flatMap {
+            _.tailRecM { l0Peers =>
+              l0Peers.headOption
+                .fold(
+                  NoPeersWithMajorityHash.raiseError[F, Either[SortedSet[L0Peer], LatestSnapshotTuple]]
+                ) { l0Peer =>
+                  (for {
+                    msd <- getLatestMajoritySnapshotData(majorityPeerIds)
+                    (snapshot, info) <- pullLatestSnapshotFromPeer(l0Peer)
```

Comment:
For me it seems like this pull should happen before the majority data calculation - we are discarding the snapshot if it's above the majority ordinal, and by pulling after we increase the probability of that happening.

I would also think about determining only the majority ordinal first and then hash in the verify function. One property that acts against it is that we do have the suppression of non majority outcome for determining the hash but we don't have it for the ordinal, so maybe we could leave it as it is right now. Why I would change it if the ordinal calculation had that property is that the computational complexity would be lessened - we would only calculate ordinal and hash once, whereas with majority snapshot data being calculated we may be calculating the hash twice in the single run of this function.